packages/@aws-cdk/aws-codepipeline/lib/action.ts

```diff
@@ -82,7 +96,108 @@ export interface ActionProperties {
    * The service provider that the action calls.
    */
   readonly provider: string;
+
+  /**
+   * The owner of this provider
+   */
+  readonly owner: string;
+
+  /**
+   * The version of this provider
+   */
+  readonly version: string;
+
+  /**
+   * The order in which AWS CodePipeline runs this action.
+   * For more information, see the AWS CodePipeline User Guide.
+   *
+   * https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html#action-requirements
+   */
+  readonly runOrder: number;
+
+  /**
+   * Artifact bounds that apply to this artifact
+   */
+  readonly artifactBounds: ActionArtifactBounds;
+
+  /**
+   * Input artifacts
+   */
+  readonly inputs: Artifact[];
+
+  /**
+   * Output artifacts
+   */
+  readonly outputs: Artifact[];
+
+  /**
+   * The name of the namespace to use for variables emitted by this action.
+   *
+   * @default - a name will be generated, based on the stage and action names
+   */
+  readonly variablesNamespace?: string;
+}
+
+/**
+ * Configuration properties for an Action
+ */
+export interface ActionProps {
+  readonly actionName: string;
+  readonly role?: iam.IRole;
+
+  /**
+   * The AWS region the given Action executes in.
+   * Note that a cross-region Pipeline requires replication buckets to function correctly.
+   * You can provide their names with the {@link PipelineProps#crossRegionReplicationBuckets} property.
+   * If you don't, the CodePipeline Construct will create new Stacks in your CDK app containing those buckets,
+   * that you will need to `cdk deploy` before deploying the main, Pipeline-containing Stack.
+   *
+   * @default the Action resides in the same region as the Pipeline
+   */
+  readonly region?: string;
+
+  /**
+   * The account the Action is supposed to execute in.
+   * For Actions backed by resources,
+   * this is inferred from the Stack {@link resource} is part of.
+   * However, some Actions, like the CloudFormation ones,
+   * are not backed by any resource, and they still might want to be cross-account.
+   * In general, a concrete Action class should specify either {@link resource},
+   * or {@link account} - but not both.
+   */
+  readonly account?: string;
+
+  /**
+   * The optional resource that is backing this Action.
+   * This is used for automatically handling Actions backed by
+   * resources from a different account and/or region.
+   */
+  readonly resource?: IResource;
+
+  /**
+   * The category of the action.
+   * The category defines which action type the owner
+   * (the entity that performs the action) performs.
+   */
+  readonly category: ActionCategory;
+
+  /**
+   * The service provider that the action calls.
+   */
+  readonly provider: string;
+
+  /**
+   * The owner of this provider
+   *
+   * @default 'AWS'
+   */
   readonly owner?: string;
+
+  /**
+   * The version of this provider
+   *
+   * @default '1'
+   */
   readonly version?: string;
```

Comment:
> Make ActionProperties strictly for output and ActionProps
> strictly for input, and put the Action in charge of defaults (instead
> of every consumer).

The Action class should be responsible for the defaulting, not the consumer. It was super annoying to work with to put things like `action.actionProperties.inputs ?? []` in every place we read them, especially since they never will actually be `undefined` (but the type signature can't show that).

Reply:
Defaulting of these values should be done not in an Action class, but in the Pipeline construct. This is part of treating all `IAction` implementations equally.

Are the changes you made to `ActionProperties` backwards compatible?